File: frogment/src/main/java/ch/grze/frogment/core/backstack/BackStackFrogmentManager.java
package ch.grze.frogment.core.backstack;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import java.util.List;

import ch.grze.frogment.frogment.FrogmentInterface;

public class BackStackFrogmentManager implements FragmentManager.OnBackStackChangedListener {
    private final BackStackChangeListener backStackChangeListener;
    private final FragmentManager fragmentManager;
    private int lastBackStackEntryCount = 0;

    public BackStackFrogmentManager(FragmentManager fragmentManager, BackStackChangeListener backStackChangeListener) {
        this.fragmentManager = fragmentManager;
        this.backStackChangeListener = backStackChangeListener;
        this.lastBackStackEntryCount = fragmentManager.getBackStackEntryCount();

        fragmentManager.addOnBackStackChangedListener(this);
    }

    @Override
    public void onBackStackChanged() {
        final int currentBackStackEntryCount = fragmentManager.getBackStackEntryCount();

        if (haveFragments()) {
            final FrogmentInterface frogment = getParentFragment();

            if (frogment != null) {
                if (wasPushed(currentBackStackEntryCount)) {
                    backStackChangeListener.onFrogmentPushed(frogment);
                }

                if (wasPopped(currentBackStackEntryCount)) {
                    backStackChangeListener.onFrogmentPopped(frogment);
                }
            }
        } else {
            backStackChangeListener.onBackStackEmpty();
        }

        lastBackStackEntryCount = currentBackStackEntryCount;
    }

    private boolean wasPushed(int backStackEntryCount) {
        return lastBackStackEntryCount < backStackEntryCount;
    }

    private boolean wasPopped(int backStackEntryCount) {
        return lastBackStackEntryCount > backStackEntryCount;
    }

    private boolean haveFragments() {
        final List<Fragment> fragmentList = fragmentManager.getFragments();

        return fragmentList != null && !fragmentList.isEmpty();
    }

    private FrogmentInterface getParentFragment() {
        final List<Fragment> fragmentList = fragmentManager.getFragments();

        return (FrogmentInterface) fragmentList.get(Math.max(0, fragmentList.size() - 2));
    }
}

File: frogment/src/main/java/ch/grze/frogment/core/backstack/BackStackFrogmentManager.kt
package ch.grze.frogment.core.backstack

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import ch.grze.frogment.frogment.FrogmentInterface


class BackStackFrogmentManager(private val fragmentManager: FragmentManager,
                               private val listener: BackStackChangeListener) : FragmentManager.OnBackStackChangedListener {

    private var lastBackStackEntryCount: Int = 0

    init {
        lastBackStackEntryCount = fragmentManager.backStackEntryCount

        fragmentManager.addOnBackStackChangedListener(this)
    }

    override fun onBackStackChanged() {
        val backStackEntryCount = fragmentManager.backStackEntryCount
        val fragments = fragmentManager.fragments

        if (hasFragments(fragments)) {
            getParentFrogment(fragments)?.let {
                if (wasPushed(backStackEntryCount)) {
                    listener.onFrogmentPushed(it)
                }

                if (wasPopped(backStackEntryCount)) {
                    listener.onFrogmentPopped(it)
                }
            }
        } else {
            listener.onBackStackEmpty()
        }

        lastBackStackEntryCount = backStackEntryCount
    }

    private fun wasPushed(backStackEntryCount: Int) = lastBackStackEntryCount < backStackEntryCount

    private fun wasPopped(backStackEntryCount: Int) = lastBackStackEntryCount > backStackEntryCount

    private fun hasFragments(fragments: List<Fragment>?) = fragments?.isEmpty()?.not() ?: false

    private fun getParentFrogment(fragments: List<Fragment>) =
        fragments[Math.max(0, fragments.size - 2)] as? FrogmentInterface
}